core-api/src/main/java/com/optimizely/ab/bucketing/DecisionService.java

```diff
@@ -198,94 +233,133 @@ public FeatureDecision getVariationForFeature(@Nonnull FeatureFlag featureFlag,
+    Variation getWhitelistedVariation(@Nonnull Experiment experiment,
+                                      @Nonnull String userId,
+                                      @Nonnull List<OptimizelyDecideOption> options,
+                                      @Nonnull DecisionReasons reasons) {
         // if a user has a forced variation mapping, return the respective variation
         Map<String, String> userIdToVariationKeyMap = experiment.getUserIdToVariationKeyMap();
         if (userIdToVariationKeyMap.containsKey(userId)) {
             String forcedVariationKey = userIdToVariationKeyMap.get(userId);
             Variation forcedVariation = experiment.getVariationKeyToVariationMap().get(forcedVariationKey);
             if (forcedVariation != null) {
-                logger.info("User \"{}\" is forced in variation \"{}\".", userId, forcedVariationKey);
+                String message = reasons.addInfo("User \"%s\" is forced in variation \"%s\".", userId, forcedVariationKey);
+                logger.info(message);
             } else {
-                logger.error("Variation \"{}\" is not in the datafile. Not activating user \"{}\".",
+                String message = reasons.addInfo("Variation \"%s\" is not in the datafile. Not activating user \"%s\".",
                     forcedVariationKey, userId);
+                logger.error(message);
             }
             return forcedVariation;
         }
         return null;
     }
 
+    @Nullable
+    Variation getWhitelistedVariation(@Nonnull Experiment experiment,
+                                      @Nonnull String userId) {
+
+        return getWhitelistedVariation(experiment, userId, Collections.emptyList(), new DecisionReasons());
+    }
+
     /**
      * Get the {@link Variation} that has been stored for the user in the {@link UserProfileService} implementation.
      *
      * @param experiment  {@link Experiment} in which the user was bucketed.
      * @param userProfile {@link UserProfile} of the user.
+     * @param projectConfig      The current projectConfig
+     * @param options            An array of decision options
+     * @param reasons            Decision log messages
      * @return null if the {@link UserProfileService} implementation is null or the user was not previously bucketed.
      * else return the {@link Variation} the user was previously bucketed into.
      */
     @Nullable
     Variation getStoredVariation(@Nonnull Experiment experiment,
                                  @Nonnull UserProfile userProfile,
-                                 @Nonnull ProjectConfig projectConfig) {
+                                 @Nonnull ProjectConfig projectConfig,
+                                 @Nonnull List<OptimizelyDecideOption> options,
+                                 @Nonnull DecisionReasons reasons) {
+
+        if (options.contains(OptimizelyDecideOption.IGNORE_USER_PROFILE_SERVICE)) return null;
```

Comment:
Can't we make all this Variation method independent of options? And from the root level, we should decide?